File: backend/app/schemas/jianghu.py
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional, List
from enum import Enum


class PracticeType(str, Enum):
    APPLIED = "applied"
    TAUGHT = "taught"
    ITERATED = "iterated"
    FAILED = "failed"
    OBSERVED = "observed"


class PracticeRecordCreate(BaseModel):
    target_type: str = Field(..., pattern=r"^(note|knowledge_unit)$", description="Target content type")
    target_id: str = Field(..., description="Target note or knowledge unit ID")
    practice_type: PracticeType = Field(..., description="Type of practice")
    description: str = Field(..., min_length=1, max_length=5000, description="What was done")
    result: Optional[str] = Field(None, max_length=5000, description="Outcome")
    learned_lesson: Optional[str] = Field(None, max_length=5000, description="Key lesson learned")
    context_snapshot: Optional[str] = Field(None, max_length=5000, description="Context snapshot as JSON string")


class PracticeRecordResponse(BaseModel):
    id: str = Field(..., description="Practice record ID")
    user_id: str = Field(..., description="Owner user ID")
    target_type: str = Field(..., description="Target content type")
    target_id: str = Field(..., description="Target ID")
    practice_type: str = Field(..., description="Type of practice")
    description: str = Field(..., description="What was done")
    result: Optional[str] = Field(None, description="Outcome")
    learned_lesson: Optional[str] = Field(None, description="Key lesson learned")
    context_snapshot: Optional[str] = Field(None, description="Context snapshot")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True


class DailyReviewGenerateRequest(BaseModel):
    review_date: Optional[date] = Field(None, description="Date to review, defaults to today")
    include_attention: bool = Field(True, description="Include attention data")
    include_notes: bool = Field(True, description="Include notes created today")
    include_knowledge: bool = Field(True, description="Include knowledge operations today")
    brain_side: Optional[str] = Field("both", description="Filter by brain side: personal / network / both")
    preferred_model: Optional[str] = Field(None, description="Preferred LLM model identifier")


class DailyReviewResponse(BaseModel):
    id: str = Field(..., description="Daily review ID")
    user_id: str = Field(..., description="Owner user ID")
    review_date: date = Field(..., description="Review date")
    content_summary: Optional[str] = Field(None, description="Summary of today's content")
    ai_reflection: Optional[str] = Field(None, description="AI reflection")
    gaps_found: List[str] = Field(default_factory=list, description="Gaps between intent and behavior")
    action_items: List[str] = Field(default_factory=list, description="Actionable improvements")
    praise_items: List[str] = Field(default_factory=list, description="Things done well")
    status: str = Field("pending", description="Status: pending / generated / reviewed / archived")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True


class DailyReviewUpdate(BaseModel):
    status: Optional[str] = Field(None, description="Status: pending / generated / reviewed / archived")
    content_summary: Optional[str] = None
    ai_reflection: Optional[str] = None
    gaps_found: Optional[List[str]] = None
    action_items: Optional[List[str]] = None
    praise_items: Optional[List[str]] = None


class RelevanceCheckRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=50000, description="External knowledge content")
    content_type: Optional[str] = Field("book_excerpt", description="Type of external content")
    user_context_summary: Optional[str] = Field(None, max_length=2000, description="User context summary")
    brain_side: Optional[str] = Field("both", description="Filter context by brain side: personal / network / both")
    preferred_model: Optional[str] = Field(None, description="Preferred LLM model identifier")


class RelevanceCheckResponse(BaseModel):
    personal_relevance_score: float = Field(..., ge=0, le=1, description="Relevance score 0-1")
    reason: str = Field(..., description="Reasoning")
    connection_evidence: Optional[str] = Field(None, description="Evidence of connection to user interests")
    first_action: Optional[str] = Field(None, description="Suggested first practical action")
    suggested_action: str = Field(..., description="Recommendation: import / import_with_practice / read_later / ignore")


class EvolutionDistribution(BaseModel):
    collected: int = Field(0)
    understood: int = Field(0)
    practiced: int = Field(0)
    validated: int = Field(0)
    internalized: int = Field(0)


class KnowledgeHealthResponse(BaseModel):
    total_items: int = Field(0, description="Total note + knowledge unit count")
    evolution_distribution: EvolutionDistribution = Field(default_factory=EvolutionDistribution)
    avg_practice_depth: float = Field(0.0)
    avg_invoke_count: float = Field(0.0)
    high_value_items: int = Field(0, description="Items with practice_depth >= 3 and invoke_count >= 5")
    zombie_items: int = Field(0, description="Items with invoke_count == 0 and age > 30 days")
    daily_active_rate: float = Field(0.0, description="Ratio of items invoked today")
    value_score_total: float = Field(0.0, description="Sum of value scores")
    health_score: float = Field(0.0, description="Composite health score 0-100: 0.5*active + 0.3*practiced + 0.2*high_value ratios")


class ContextGuideScope(str, Enum):
    PERSONAL = "personal"
    NETWORK = "network"
    BOTH = "both"


class ContextGuideBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200, description="Guide title")
    content: str = Field(..., min_length=1, max_length=50000, description="Markdown guide content")
    scope: ContextGuideScope = Field(ContextGuideScope.BOTH, description="Applicable brain side")
    is_active: bool = Field(True, description="Whether this guide is active")
    version_tag: Optional[str] = Field(None, max_length=50, description="Optional version tag")


class ContextGuideCreate(ContextGuideBase):
    pass


class ContextGuideUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = Field(None, min_length=1, max_length=50000)
    scope: Optional[ContextGuideScope] = None
    is_active: Optional[bool] = None
    version_tag: Optional[str] = Field(None, max_length=50)


class ContextGuideResponse(ContextGuideBase):
    id: str = Field(..., description="Guide ID")
    user_id: str = Field(..., description="Owner user ID")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True


class ContextGuideGenerateRequest(BaseModel):
    brain_side: Optional[str] = Field("both", description="Filter content by brain side: personal / network / both")
    preferred_model: Optional[str] = Field(None, description="Preferred LLM model identifier")
    title: Optional[str] = Field(None, max_length=200, description="Optional title for generated guide")


class CognitivePotentialItem(BaseModel):
    content_id: str = Field(..., description="Source note or knowledge unit ID")
    content_type: str = Field(..., description="note / knowledge_unit")
    title: str = Field(..., description="Short title of the content")
    score: float = Field(..., ge=0, le=1, description="Potential score 0-1")
    reason: str = Field(..., description="Why this content has this kind of potential")
    suggested_action: str = Field(..., description="Concrete next step")


class CognitivePotentialResponse(BaseModel):
    summary: str = Field(..., description="Overall assessment")
    sinkable: List[CognitivePotentialItem] = Field(default_factory=list, description="Items that can be internalized into habits")
    outputable: List[CognitivePotentialItem] = Field(default_factory=list, description="Items that can be turned into articles, courses, shares")
    monetizable: List[CognitivePotentialItem] = Field(default_factory=list, description="Items with market value")


class CognitivePotentialRequest(BaseModel):
    brain_side: Optional[str] = Field("both", description="Filter content by brain side: personal / network / both")
    preferred_model: Optional[str] = Field(None, description="Preferred LLM model identifier")


class ExperimentStatus(str, Enum):
    PLANNED = "planned"
    RUNNING = "running"
    COMPLETED = "completed"
    ABANDONED = "abandoned"


class ExperimentLogBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200, description="Experiment title")
    hypothesis: str = Field(..., min_length=1, max_length=5000, description="What you assume")
    controlled_variable: Optional[str] = Field(None, max_length=2000, description="The single variable being tested")
    expected_result: Optional[str] = Field(None, max_length=2000, description="Expected outcome")
    actual_result: Optional[str] = Field(None, max_length=5000, description="Actual observed outcome")
    conclusion: Optional[str] = Field(None, max_length=5000, description="Conclusion or learning")
    status: ExperimentStatus = Field(ExperimentStatus.PLANNED, description="Experiment status")
    related_content_type: Optional[str] = Field(None, pattern=r"^(note|knowledge_unit)$", description="Linked content type")
    related_content_id: Optional[str] = Field(None, description="Linked content ID")
    brain_side: Optional[str] = Field("both", description="Brain side context")


class ExperimentLogCreate(ExperimentLogBase):
    pass


class ExperimentLogUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    hypothesis: Optional[str] = Field(None, min_length=1, max_length=5000)
    controlled_variable: Optional[str] = Field(None, max_length=2000)
    expected_result: Optional[str] = Field(None, max_length=2000)
    actual_result: Optional[str] = Field(None, max_length=5000)
    conclusion: Optional[str] = Field(None, max_length=5000)
    status: Optional[ExperimentStatus] = None
    related_content_type: Optional[str] = Field(None, pattern=r"^(note|knowledge_unit)$")
    related_content_id: Optional[str] = None
    brain_side: Optional[str] = None


class ExperimentLogResponse(ExperimentLogBase):
    id: str = Field(..., description="Experiment log ID")
    user_id: str = Field(..., description="Owner user ID")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True

File: frontend/src/api/jianghu.ts
import api from './client';

export interface PracticeRecord {
  id: string;
  user_id: string;
  target_type: 'note' | 'knowledge_unit';
  target_id: string;
  practice_type: 'applied' | 'taught' | 'iterated' | 'failed' | 'observed';
  description: string;
  result?: string;
  learned_lesson?: string;
  context_snapshot?: string;
  created_at: string;
  updated_at: string;
}

export interface PracticeRecordCreateData {
  target_type: 'note' | 'knowledge_unit';
  target_id: string;
  practice_type: 'applied' | 'taught' | 'iterated' | 'failed' | 'observed';
  description: string;
  result?: string;
  learned_lesson?: string;
  context_snapshot?: string;
}

export interface DailyReview {
  id: string;
  user_id: string;
  review_date: string;
  content_summary?: string;
  ai_reflection?: string;
  gaps_found: string[];
  action_items: string[];
  praise_items: string[];
  status: 'pending' | 'generated' | 'reviewed' | 'archived';
  created_at: string;
  updated_at: string;
}

export interface RelevanceCheckRequest {
  content: string;
  content_type?: string;
  user_context_summary?: string;
  brain_side?: string;
}

export interface RelevanceCheckResponse {
  personal_relevance_score: number;
  reason: string;
  connection_evidence?: string;
  first_action?: string;
  suggested_action: 'import' | 'import_with_practice' | 'read_later' | 'ignore';
}

export interface EvolutionDistribution {
  collected: number;
  understood: number;
  practiced: number;
  validated: number;
  internalized: number;
}

export interface KnowledgeHealthResponse {
  total_items: number;
  evolution_distribution: EvolutionDistribution;
  avg_practice_depth: number;
  avg_invoke_count: number;
  high_value_items: number;
  zombie_items: number;
  daily_active_rate: number;
  value_score_total: number;
  health_score: number;
}

export interface ContextGuide {
  id: string;
  user_id: string;
  title: string;
  content: string;
  scope: 'personal' | 'network' | 'both';
  is_active: boolean;
  version_tag?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ContextGuideCreateData {
  title: string;
  content: string;
  scope?: 'personal' | 'network' | 'both';
  is_active?: boolean;
  version_tag?: string;
}

export interface ContextGuideUpdateData {
  title?: string;
  content?: string;
  scope?: 'personal' | 'network' | 'both';
  is_active?: boolean;
  version_tag?: string;
}

export interface CognitivePotentialItem {
  content_id: string;
  content_type: 'note' | 'knowledge_unit';
  title: string;
  score: number;
  reason: string;
  suggested_action: string;
}

export interface CognitivePotentialResponse {
  summary: string;
  sinkable: CognitivePotentialItem[];
  outputable: CognitivePotentialItem[];
  monetizable: CognitivePotentialItem[];
  analyzed_at?: string | null;
  model_used?: string | null;
}

export interface ExperimentLog {
  id: string;
  user_id: string;
  title: string;
  hypothesis: string;
  controlled_variable?: string | null;
  expected_result?: string | null;
  actual_result?: string | null;
  conclusion?: string | null;
  status: 'planned' | 'running' | 'completed' | 'abandoned';
  related_content_type?: 'note' | 'knowledge_unit' | null;
  related_content_id?: string | null;
  brain_side: string;
  created_at: string;
  updated_at: string;
}

export interface ExperimentLogCreateData {
  title: string;
  hypothesis: string;
  controlled_variable?: string;
  expected_result?: string;
  actual_result?: string;
  conclusion?: string;
  status?: 'planned' | 'running' | 'completed' | 'abandoned';
  related_content_type?: 'note' | 'knowledge_unit' | null;
  related_content_id?: string | null;
  brain_side?: string;
}

export interface ExperimentLogUpdateData {
  title?: string;
  hypothesis?: string;
  controlled_variable?: string;
  expected_result?: string;
  actual_result?: string;
  conclusion?: string;
  status?: 'planned' | 'running' | 'completed' | 'abandoned';
  related_content_type?: 'note' | 'knowledge_unit' | null;
  related_content_id?: string | null;
  brain_side?: string;
}

export interface EvolutionTransition {
  id: string;
  content_type: 'note' | 'knowledge_unit';
  content_id: string;
  // 内容已删除时后端返回 null
  title: string | null;
  from_stage: string;
  to_stage: string;
  trigger: 'practice' | 'manual';
  created_at: string;
}

export const jianghuApi = {
  // Practice records
  listPracticeRecords: (params?: { target_type?: string; target_id?: string; practice_type?: string; brain_side?: string; limit?: number; offset?: number }) =>
    api.get<PracticeRecord[]>('/api/v1/jianghu/practice-records', { params }),
  createPracticeRecord: (data: PracticeRecordCreateData) =>
    api.post<PracticeRecord>('/api/v1/jianghu/practice-records', data),
  getPracticeRecord: (id: string) => api.get<PracticeRecord>(`/api/v1/jianghu/practice-records/${id}`),
  updatePracticeRecord: (id: string, data: PracticeRecordCreateData) =>
    api.put<PracticeRecord>(`/api/v1/jianghu/practice-records/${id}`, data),
  deletePracticeRecord: (id: string) => api.delete(`/api/v1/jianghu/practice-records/${id}`),

  // Daily reviews
  generateDailyReview: (
    data?: { review_date?: string; include_attention?: boolean; include_notes?: boolean; include_knowledge?: boolean; brain_side?: string },
    preferred_model?: string
  ) =>
    api.post<DailyReview>('/api/v1/jianghu/daily-reviews/generate', { ...(data || {}), preferred_model }),
  listDailyReviews: (params?: { status?: string; limit?: number; offset?: number }) =>
    api.get<DailyReview[]>('/api/v1/jianghu/daily-reviews', { params }),
  getDailyReview: (id: string) => api.get<DailyReview>(`/api/v1/jianghu/daily-reviews/${id}`),
  updateDailyReview: (id: string, data: Partial<DailyReview>) =>
    api.put<DailyReview>(`/api/v1/jianghu/daily-reviews/${id}`, data),

  // Relevance & health
  checkRelevance: (data: RelevanceCheckRequest, preferred_model?: string) =>
    api.post<RelevanceCheckResponse>('/api/v1/jianghu/relevance-check', { ...data, preferred_model }),
  getKnowledgeHealth: (brain_side?: string) => api.get<KnowledgeHealthResponse>('/api/v1/jianghu/knowledge-health', { params: { brain_side } }),
  // 最近的进化阶段跃迁记录（知识自进化线）
  listEvolutionTransitions: (limit = 50) =>
    api.get<EvolutionTransition[]>('/api/v1/jianghu/evolution-transitions', { params: { limit } }),

  // Context guides
  listContextGuides: (is_active?: boolean) =>
    api.get<ContextGuide[]>('/api/v1/jianghu/context-guides', { params: is_active !== undefined ? { is_active } : undefined }),
  createContextGuide: (data: ContextGuideCreateData) =>
    api.post<ContextGuide>('/api/v1/jianghu/context-guides', data),
  getContextGuide: (id: string) => api.get<ContextGuide>(`/api/v1/jianghu/context-guides/${id}`),
  updateContextGuide: (id: string, data: ContextGuideUpdateData) =>
    api.put<ContextGuide>(`/api/v1/jianghu/context-guides/${id}`, data),
  deleteContextGuide: (id: string) => api.delete(`/api/v1/jianghu/context-guides/${id}`),
  generateContextGuide: (data?: { brain_side?: string; preferred_model?: string; title?: string }) =>
    api.post<ContextGuide>('/api/v1/jianghu/context-guides/generate', data || {}),
  analyzeCognitivePotential: (data?: { brain_side?: string; preferred_model?: string }) =>
    api.post<CognitivePotentialResponse>('/api/v1/jianghu/cognitive-potential', data || {}),
  // 最近一次已保存的分析（免费读；404=还没有结果，由调用方归一为 null）
  getCognitivePotentialLatest: (brainSide: string) =>
    api.get<CognitivePotentialResponse>('/api/v1/jianghu/cognitive-potential/latest', { params: { brain_side: brainSide } })
      .then((r) => r.data)
      .catch((err: any) => {
        if (err?.status === 404) return null;
        throw err;
      }),

  // Experiment logs
  listExperimentLogs: (params?: { status?: string; brain_side?: string }) =>
    api.get<ExperimentLog[]>('/api/v1/jianghu/experiment-logs', { params }),
  createExperimentLog: (data: ExperimentLogCreateData) =>
    api.post<ExperimentLog>('/api/v1/jianghu/experiment-logs', data),
  getExperimentLog: (id: string) => api.get<ExperimentLog>(`/api/v1/jianghu/experiment-logs/${id}`),
  updateExperimentLog: (id: string, data: ExperimentLogUpdateData) =>
    api.put<ExperimentLog>(`/api/v1/jianghu/experiment-logs/${id}`, data),
  deleteExperimentLog: (id: string) => api.delete(`/api/v1/jianghu/experiment-logs/${id}`),
};
